### DISCUSSION TITLE
Can you use `test` and `include` together in overrides?

### DISCUSSION THREAD

**Discussion by dietergeerts** [babel/babel#13988] [Q&A]
Hi, I was making some sharable configurations, and as I'm used to using the test and include options of Webpack together, I thought of doing the same in the babel config, like this:

```
module.exports = /** @type {babel.TransformOptions} */ ({
    overrides: [{
        test: /\.ts$/,
        include: webapp.patterns.sources.all.regExp,
        presets: [
            [require.resolve('@babel/preset-env'), {targets: {node: 'current'}}],
            [require.resolve('@babel/preset-typescript')]
        ]
    }]
})
```

The way I see this is that `test` is for identifying the type of files, while `include` tells where these files can be located.

But then I saw the following in the docs:

> This option is a synonym for "test".

So does this mean that only the `include` option in my config will be considered? Or does this work like in Webpack, where both will be considered? If they are both considered, it would be better to update the docs to state this, as then probably the current docs are actually meaning that you can use `include` alone to, no?

**Comment by JLHwung** [comment]
See

https://github.com/babel/babel/blob/a470f7b479c649fcaea15df473dce395f4786731/packages/babel-core/src/config/config-chain.ts#L757-L770

for how Babel handles `test` and `include` in config items. In your case,

```
{
  test: /\.ts$/,
  include: webapp.patterns.sources.all.regExp,
}
```
is equivalent to
```js
{
  test: [/\.ts$/, webapp.patterns.sources.all.regExp]
}
```

> If they are both considered, it would be better to update the docs to state this

Can you open a PR to https://github.com/babel/website/blob/main/docs/options.md#include?

**Comment by dietergeerts** [comment]
🤔 , but that does mean that there is no way to do the following: when the file is a typescript file, and it's a source file from packages from our own scope. So there is no way to define an "AND" statement? Or am I overlooking something here?

I am aware that I can just combine them into 1, but I believe it's confusing because how similar this config is to webpack, and that webpack handles this differently. From their config examples:

```
const path = require('path');

module.exports = {
  //...
  module: {
    rules: [
      {
        test: /\\.css$/,
        include: [
          // will include any paths relative to the current directory starting with `app/styles`
          // e.g. `app/styles.css`, `app/styles/styles.css`, `app/stylesheet.css`
          path.resolve(__dirname, 'app/styles'),
          // add an extra slash to only include the content of the directory `vendor/styles/`
          path.join(__dirname, 'vendor/styles/'),
        ],
      },
    ],
  },
};
```
Here you see that they are like an AND statement, where both `test` and `include` has to pass, while in Babel, you are now stating that this is not the case, as both are combined and then only 1 of them has to succeed.

I believe that this will be a common mistake in many configuration, just because one can assume it works like webpack does, and the babel docs aren't really pointing out how they actually work. Off course, they state that include is a synonym of test, but what does that mean when they are both used?

**Comment by dietergeerts** [comment]
@JLHwung , but after reading your link again, are you sure that they are equivalent? If I apply the following:
```
(options.test === undefined || 
       configFieldIsApplicable(context, options.test, dirname)) && 
     (options.include === undefined || 
       configFieldIsApplicable(context, options.include, dirname)) && 
     (options.exclude === undefined || 
       !configFieldIsApplicable(context, options.exclude, dirname)) 
```
on:
```
{
  test: /\.ts$/,
  include: webapp.patterns.sources.all.regExp,
}
```
Then I get for a `ts` file inside a location other than my include:

- `(false || true) && (false || false) && (true || ...)`
- `true && false && true`
- `false`

applied on:
```
{
  test: [/\.ts$/, webapp.patterns.sources.all.regExp]
}
```
- `(false || true) && (true || ...) && (true || ...)`
- `true && true && true`
- `true`

Or am I really going crazy here?

So I can see why the docs state that it's a synonym, because you can either use `test` or `include`. But they can also be used together in the way that webpack does it, the way that I was expecting it to work, where they behave like AND `test` must succeed AND `include` must be the case. Which is great, but the docs can be improved to clearly show this. Even the webapp docs aren't really clear on this too though.

**Reply by JLHwung** [reply]
My bad. You are right. The `test` and `includes` behave as an AND statement while multiple `test` options are concatenated with OR. So the examples I mentioned above are not equivalent. Thank you for clarifying that!
